libparsec/crates/client/src/certif/shamir_recovery_setup.rs

```diff
@@ -0,0 +1,363 @@
+    }
+
+    // 3. Check organization's and author's status
+
+    // 4. Check recipients status
+
+    let mut participants_id: HashSet<_> = share_recipients.keys().collect();
+    participants_id.insert(&author_id);
+    // TODO implement a get_users method instead of list_user (let the DB do its job)
+    let participants = certificate_ops.list_users(true, None, None).await?;
+    let participants = participants
+        .iter()
+        .filter(|info| participants_id.contains(&info.id));
+
+    // no participant is missing
+    for &&id in &participants_id {
+        if !participants.clone().any(|info| info.id == id) {
+            return Err(CertifShamirError::MissingUser(id));
+        }
+    }
+
+    // all participants not revoked
+    for info in participants {
+        if info.revoked_on.is_some() {
+            return Err(CertifShamirError::UserRevoked(info.id));
+        }
+    }
+    // 5. Generate certificates
+
+    let brief = ShamirRecoveryBriefCertificate {
+        author: author_device,
+        timestamp,
+        user_id: author_id,
+        threshold: NonZeroU64::new(threshold.into()).expect("shamir threshold is zero"), // checked during the first step
+        per_recipient_shares: share_recipients
+            .iter()
+            .map(|(&k, &v)| (k, NonZeroU64::new(v.into()).expect("Share count is zero")))
+            .collect(),
+    }
+    .dump_and_sign(&certificate_ops.device.signing_key)
+    .into();
+
+    let recovery_device = LocalDevice::generate_new_device(
+        certificate_ops.device.organization_addr.clone(),
+        certificate_ops.device.initial_profile,
+        certificate_ops.device.human_handle.clone(),
+        DeviceLabel::default(),
+        Some(certificate_ops.device.user_id),
+        None,
+        None,
+        None,
+    );
+
+    let data_key = SecretKey::generate();
+    let reveal_token = InvitationToken::default();
+
+    let ciphered_data = data_key.encrypt(&recovery_device.dump()).into();
+
+    let shark_shares = shamir_make_shares(
+        threshold,
+        &ShamirRecoverySecret {
+            data_key,
+            reveal_token,
+        }
+        .dump()?,
+        total_share_count.into(),
+    );
+    let mut idx = 0_usize;
+    let shares = share_recipients
+        .iter()
+        .map(|(&share_recipient_id, &share_count)| {
+            let shares: Vec<u8> = shark_shares[idx..idx + share_count as usize].iter().fold(
+                vec![],
+                |mut acc, share| {
+                    acc.append(&mut Vec::<u8>::from(share));
+                    acc
+                },
+            );
+            idx += share_count as usize;
+            ShamirRecoveryShareCertificate {
+                author: author_device,
+                timestamp,
+                user_id: author_id,
+                recipient: share_recipient_id,
+                ciphered_share: certificate_ops.device.signing_key.sign(&shares),
```

Comment:
this share should be ciphered by the recipient public key (and not signed by the author signing key)